src/FSharp.Data.GraphQL.Client/Subscriptions.fs

```diff
@@ -0,0 +1,247 @@
+            match r |> Array.tryFind (fun (name, _) -> name.ToLowerInvariant() = "id") with
+            | Some (_, JsonValue.String id) -> id.ToLowerInvariant()
+            | _ -> failwith "Expected server to return a valid message, but message ID field is missing in response, or the ID field is not a string field."
+        match JsonValue.Parse(s) with
+        | JsonValue.Record r ->
+            match getType r with
+            | "connection_ack" -> ConnectionAck
+            | "keep_alive" -> KeepAlive
+            | "connection_error" -> ConnectionError (getPayload r)
+            | "data" -> Data (getId r, getPayload r)
+            | "error" -> Error (getId r, getPayload r)
+            | "complete" -> Complete (getId r)
+            | other -> failwithf "Unexpected message type received from server. Can not treat \"%s\" message type." other
+        | other -> failwithf "Unexpected response JSON type. Expected a Record Json, but got %A." other
+
+/// A subscription response message, following the protocol of the handler that will use it.
+type IGraphQLSubscriptionResponse =
+    inherit IDisposable
+    /// When implemented, returns the immediate data for the operation.
+    abstract Data : JsonValue option
+    /// When implemented, returns the deferred data (@defer, @stream and subscriptions) for the operation.
+    abstract Deferred : IObservable<JsonValue>
+    /// When implemented, unsubscribe from the operation using the handler of the protocol.
+    abstract Unsubscribe : unit -> unit
+    /// When implemented, asynchronously unsubscribe from the operation using the handler of the protocol.
+    abstract AsyncUnsubscribe : unit -> Async<unit>
+
+type internal SubscriptionResponseMessage =
+    | Incoming of payload : JsonValue
+    | Finished
+
+/// The default GraphQL subscription response implementation that uses GraphQL Over Web Socket protocol.
+/// See https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md for information.
+type GraphQLSubscriptionResponse internal (source : IObservable<SubscriptionResponseMessage>, unsubscriber : unit -> Async<unit>) =
+    let tryPickFirstAndKeepRemaining (source : IObservable<'T>) =
+        let sourceSeq = Observable.toSeq source
+        let first = Seq.tryHead sourceSeq
+        match first with
+        | Some first -> Some (first, Seq.skip 1 sourceSeq)
+        | None -> None
+
+    let data, deferred =
+        source
+        |> tryPickFirstAndKeepRemaining
+        |> Option.map (fun (first, remaining) ->
+            let deferred = {
+                new IObservable<JsonValue> with
+                    member __.Subscribe(observer) =
+                        remaining
+                        |> Seq.iter (function
+                            | Incoming payload -> observer.OnNext(payload)
+                            | Finished -> observer.OnCompleted())
+                        { new IDisposable with member __.Dispose() = () } }
```

Comment:
We pick the first item from the response and use it as the direct result.
The remaining items will be processed into the deferred result as an `IObservable`.
Every Observable returns `JsonValue` here. They will be mapped into provided types in the `ProvidedTypesHelper`.